# **Lab 3: Functions and Loops (20 points)**

#### **Dylan Watson**





<br>




---

**IMPORTANT**: 
- Before starting the lab, save a copy to your google drive by navigating to `File > Save a copy in Drive` <br>
- Then rename the copied file `Lab#_LASTNAME_FIRSTNAME.ipynb` where # is the lab assignment number, LASTNAME is your last name, and FIRSTNAME is your first name.

<br>


**Instructions**: Complete the lab below.

- <font color=red>Tasks</font> indicated in red must be completed to receive credit
- Include all code for your work
- Include markdown cells with written answers explaining your work when prompted
- **NOTE: Your completed lab must be submitted to Canvas as a .ipynb file by the end of the lab session**
---

<br>

<br>






####<font color=red>Organized</font> (1 points)
* Answers should be organized in the following format for organization and readability

<br>

```
# 1. 
This will be the code block to answer the first part of a task
```

1. This will be the text answer/explanation (if prompted) for the first task

<br>

```
# 2. 
This will be the code block to answer the second part of a task in a separate code cell
```

2. This will be the text answer/explanation (if prompted) for the second task

<br>

Continue this format for the remaining subtasks


<br>

---

<br>






####<font color=red>Task 1</font>  (6 points)

1. (4 points) Create a function `quadratic_formula` that has the following attributes
* Input: Three single numbers `a`, `b`, `c` as function arguments
* Ouput: The solution to the qudratic equation
  * $\begin{align*}
\frac{-b \pm \sqrt{b^2 - 4ac}} {2a}
\end{align*}$
* Us `if` statements to potentially return three types of function outputs
  1. If $b^2 - 4ac$ is zero, only return the one solution/number
  2. If $b^2 - 4ac$ is less than zero, print a statement saying there are no zeros that are real numbers
  3. If $b^2 - 4ac$ is greater than zero, return the two real zeros as a single vector

<br>

2. (2 points) Run the following commands to confirm answers
```
quadratic_formula(1, 0, 0)
```
```
quadratic_formula(1, 0, 1)
```
```
quadratic_formula(1, 0, -1)
```


In [1]:
quadratic_formula = function(a, b, c) {
    discriminant = b^2 - 4 * a * c
    if (discriminant > 0) {
        result = (-b + c( sqrt(discriminant), -sqrt(discriminant) ) ) / (2 * a)
    } else if (discriminant < 0) {
        result = "no zeros that are real numbers"
    } else {
        result = -b / (2 * a) 
    }
    return(result)
}

print(quadratic_formula(1, 0, 0))
print(quadratic_formula(1, 0, 1))
print(quadratic_formula(1, 0, -1))

[1] 0
[1] "no zeros that are real numbers"
[1]  1 -1


First output is 0, second is no real num zeros, third is (-1, 1).

<br>

####<font color=red>Task 2</font>  (5 points)

1. (3 points) Create a function `avg_diff` that has the following attributes
  * Inputs:
    * vector1 - a numeric vector
    * vector2 - a numeric vector
  * Ouput: A list of three objects providing
    1. the difference between the two vectors
    2. the average difference between the two vectors
    3. the length of the first vector
    * `if` vectors are not the same length, do not perform the calculation. Instead, print a warning statement that the vectors are not the same length

<br>

2. (2 points) Run the following commands to confirm answers
```
vec1 <- c(1, 2, 3, 4, 5, 6)
vec2 <- c(0, 1, 2, 3, 4, 5)
avg_diff(vec1, vec2)
```
```
vec1 <- c(1, 2, 3, 4, 5, 6)
vec2 <- c(0, 1, 2, 3, 4)
avg_diff(vec1, vec2)
```


In [2]:
avg_diff = function(vector1, vector2) {
    if ( length(vector1) == length(vector2) ) {
        vector_diff = vector2 - vector1
        vector_diff_avg = mean(vector_diff)
        vector_diff_len = length(vector1)
        return( c(vector_diff, vector_diff_avg, vector_diff_len) )
    } else {
        return("Vectors are not the same length")
    }
}

vec1 <- c(1, 2, 3, 4, 5, 6)
vec2 <- c(0, 1, 2, 3, 4, 5)
print(avg_diff(vec1, vec2))
vec1 <- c(1, 2, 3, 4, 5, 6)
vec2 <- c(0, 1, 2, 3, 4)
print(avg_diff(vec1, vec2))

[1] -1 -1 -1 -1 -1 -1 -1  6
[1] "Vectors are not the same length"


First output is a vector (-1, -1, -1, -1, -1, -1) followed by an avg diff of -1 and a length of 6. Second output is vectors are not the same length.

<br>

---

<br>






####<font color=red>Task 3</font>  (5 points)

1. (1 point) From the `mtcars` dataset, find the unique values of the `gear` variable. Print.

2. (2 point) Use a `for` loop to find the average `hp` for each unique `gear`. Show code and output.

3. (1 point) Which number of gears is associated with the highest horse power `hp`? State in a complete sentence.

4. (1 point) Which cars have 5 forward gears? State in a complete sentence.


In [3]:
data(mtcars)

unique_gear = unique(mtcars$gear)
print(unique_gear)

unique_gear_hp_means = c()
for(i in unique_gear) {
    unique_gear_hp_means = c( unique_gear_hp_means, mean(mtcars$hp[mtcars$gear == i]) )
}
print(unique_gear_hp_means)

cars_five_gears = mtcars[mtcars$gear == 5, c(4, 10)]
cars_five_gears

[1] 4 3 5
[1]  89.5000 176.1333 195.6000


hp  gear
Porsche 914-2   91 5   
Lotus Europa   113 5   
Ford Pantera L 264 5   
Ferrari Dino   175 5   
Maserati Bora  335 5

The unique gears are 3, 4, and 5. The avg hp of cars with each number of gears respectively are 176.133, 89.5, and 195.6. The cars with 5 gears are the Porsche 914-2, Lotus Europa, Ford PAntera L, Ferrari Dino, and Maserati Bora.

<br>

---

<br>






####<font color=red>Task 4</font>  (3 points)

1. (1 point) Use `tapply` to find the average `hp` for each unique number of forward gears (`gear`). Show code and output.

2. (1 point) Create a dataframe with two columns. One column should be unique `gear`. One column should be average `hp`

3. (1 point) Compare to the answer for task 3.2. Are they the same? Which approach is more efficient? Explain.

In [4]:
gear_hp_means_2 = data.frame(Gears = unique(mtcars$gear), MeanHP = tapply(mtcars$hp, mtcars$gear, mean))
gear_hp_means_2

Gears MeanHP  
3 4     176.1333
4 3      89.5000
5 5     195.6000

<br>

Compared to task 3.2, the results are identical. The method of using a tapply function is significantly easier to both write and read. I would consider tapply to be more efficient.